packages/react-dom/src/client/ToStringValue.js

```diff
@@ -44,14 +45,39 @@ export opaque type TrustedValue: {toString(): string, valueOf(): string} = {
   valueOf(): string,
 };
 
+export function evaluateToStringOrTrustedType(
+  value: mixed,
+): string | TrustedValue {
+  if (
+    value !== null &&
+    typeof value === 'object' &&
+    value.$$typeof === REACT_OPAQUE_OBJECT_TYPE &&
+    typeof value._setId === 'function'
```

Comment:
You shouldn't need this extra check. You've asserted the type. Just cast it through `let obj: IdObject = (value: any)`.